### ISSUE TITLE
Creating a new metric

### ISSUE THREAD

**Issue by Illuminatian** [chaoss/grimoirelab#102]
@jgbarah 

So, over the past month or so I've been working on trying to get more familiar with the interface and making my own custom visualizations in the Kibana interface, however there's one thing I still haven't quite figured out how to do. I'm interested in creating a custom visualization, a metric. The calculating of this metric requires some simple math be done on values being pulled from the database. I can't seem to find any way to do this, can you help me out with this problem?  For specifics, I'm looking at trying to create a metric that gives the 'Pony Factor' for a repository, possibly getting it accepted into the project as a default visualization.

**Comment by jgbarah** [comment]
If you just want to compute the metric, in this specific case (the pony factor), I would compute it from the enriched indexes, since in them you have the data needed (namely, the number of commits per author).

If you want to show it in a chart in Kibiter/Kibana, you have two options:

* Compute the data over time (say, every day or every month), in a new index, and just chart that index with a Kibana visualization. This is simple, but you loss the ability of filtering and drilling down in the dashboard for  this value.

* Find a way of doing the computation from a visualization in Kibana. In this case, it is not easy, because the metric is a second-order one (I mean, it is really based in some other metrics). If you're interested in this one, we can explore together how to do it, if you want. It is likely that it could be done with TimeLion.

Of course, you can just compute the metric in, say, a Python Jupyter Notebook, and just plot it there. But you lose the interaction with the dashboard...

Please, let me know if I'm answering your question, or you were asking for something else.

**Comment by Illuminatian** [comment]
@jgbarah 

If there's a way that we could accomplish this that would make it a desirable addition to the GL project, I would prefer to do it that way, as my last assignment is to get a metric added into GL that CHAOSS would be interested in.

**Comment by Illuminatian** [comment]
@jgbarah 

Just trying to follow up on this. I think option #2 would probably be the best way to go about this, as I have no opposition to working with you to try and accomplish this. We do need to get this done within the next 2 weeks though 👍

**Comment by Illuminatian** [comment]
Hey @jgbarah, can we start working together on this soon?

**Comment by jgbarah** [comment]
Sorry @Illuminatian, I had missed your messages. Thanks @germonprez for the private heads up ;-)

Yes, please, let's work on this!

I suggest that we follow this path, if you have the time and interest:

(1) We define the metric and/or related metrics as precisely as possible (see below on this).
(2) We computer the metric, for example for git enriched indexes, in pure Python, so that we can get a gut feeling of how the metric behaves.
(3) We try to do the same in Kibana, maybe using intermediate indexes, if current enriched indexes are not good enough.

For (1), I would use a pull request on the [list of activity metrics](https://github.com/chaoss/metrics/blob/master/activity-metrics-list.md), opening a new page with the detailed description of the metric. We can use for a start the ["original" definition, by Daniel Gruno](https://ke4qqq.wordpress.com/2015/02/08/pony-factor-math/). The main problem I see with it is that it is accumulative (history is "never forgotten"), which is why I find it useful to also consider the "moving Pony Factor", which would be exactly like the Pony factor, but for a period of time. For example, the "1-year moving Pony Factor" would be computed by considering only commits during a period of 1 year. That would allow for a more fair comparison of the factor over time.

So, if you want, we can work in a pull request refining these definitions, either sticking to the original or adding the moving one, as you may prefer.

Then, we can explore (2) and (3), in that order or in parallel.

For (2), we can write some simple code that:

* Queries an enriched index for the total number of contributions (eg, commits for a git index), and contributions by contributor.
* Computes the Pony factor from that.

The basis for this is [described in the GrimoireLab tutorial](https://chaoss.github.io/grimoirelab-tutorial/python/querying-grimoireelk-indexes.html), I can help you with that.

For (3), I think we need to explore the new charting capabilities in Kibana6, since I'm not aware of how to do it with what Kibana5 provides. But I'm going to review my knowledge about that, maybe I'm missing something. If we don't find a way, we can always produce a new enriched index, that produces the needed data, to later visualize it (and contribute that code to GrimoireELK).

For practical purposes, have you already deployed GrimoireLab? Are you familiar with the analysis of repositories, and the production of enriched indexes and dashboards? If not, I can drive you through the process.

**Comment by Illuminatian** [comment]
@jgbarah 

Looks good to me, when/how would you like to collaborate on this? I have GL deployed locally through docker on my laptop, I am not familiar with how the indexes are produced though.

**Comment by Illuminatian** [comment]
I'm in Central timezone (US). Not sure how that syncs up with you, but I can be available after about 5pm my time.

I do have the docker container deployed, but it's deployed to a command line only Linux VM on my laptop, so if we need to dive into the code I don't know really how I would go about that.

**Comment by jgbarah** [comment]
> I'm in Central timezone (US). Not sure how that syncs up with you, but I can be available after about 5pm my time.

OK, I will try to be available in the grimoirelab channel at freenode IRC at about that time. Ping me there if you see me.

> I do have the docker container deployed, but it's deployed to a command line only Linux VM on my laptop, so if we need to dive into the code I don't know really how I would go about that.

You can just export the ports of the container (I think 9200 for Elasticsearch and 5601 for Kibiter will be enough). And then, write the code in your host operating system. Or we spend some time on letting you have a complete installation in your host operating system (at least of GrimoireLab, which is not difficult).

**Comment by Illuminatian** [comment]
Sounds like a plan. Talk to you then!

**Comment by Illuminatian** [comment]
@jgbarah 

I should be available around the same time today as yesterday. I had a seemingly never-ending series of issues trying to do a windows native install of GL, so after about 3 hours of banging my head against the wall, I went ahead and made a Debian VM and was able to succesfully get all the components installed in a python venv. I also go mysql installed, however when trying to run sortinghat to get it to make the needed DB, it kept throwing up errors, saying the user was not authorized, even though I had added a mysql user with all the appropriate permissions, that's about where I left off at last night.
